lib/charliecloud.py

```diff
@@ -456,6 +469,31 @@ def set_(dst_key, *src_keys):
          for k in config["config"]["Volumes"].keys():
             self.metadata["volumes"].append(k)
 
+   def metadata_append(self, data):
+      DEBUG("appending metadata: %s" % data)
+      for k in data:
+         if (k in self.metadata.keys()):
+            self.metadata[k] += data[k]
+         else:
+            FATAL("config has no key '%s'" % k)
+
+   def metadata_empty_layer_history(self, hist):
+      if (len(hist) > 0 ):
+         for i in range(len(hist)):
+            hist[i]["empty_layer"] = True
+      return hist
+
+   def metadata_history_replace(self, hist):
+      self.metadata["history"] = hist
+
+   def metadata_history_return(self):
+      "Return config history with all entires having empty_layer set to True"
+      try:
+         hist = self.metadata["history"]
+      except KeyError:
+         return []
```

Comment:
I'm seeing a lot of these. I would recommend moving the workaround for missing history to when `metadata.json` is read from disk. Coupled with the translate on pull/push, then all the internal history manipulation code can assume the data structures are consistent.